test/functional/system/rounds.js

```diff
@@ -0,0 +1,1691 @@
+					.plus(new Bignum(reward.fees))
+					.toString();
+				found.rewards = new Bignum(found.rewards)
+					.plus(new Bignum(reward.rewards))
+					.toString();
+				found.balance = new Bignum(found.balance)
+					.plus(new Bignum(reward.fees))
+					.plus(new Bignum(reward.rewards))
+					.toString();
+				found.u_balance = new Bignum(found.u_balance)
+					.plus(new Bignum(reward.fees))
+					.plus(new Bignum(reward.rewards))
+					.toString();
+			}
+		});
+
+		return accounts;
+	}
+
+	function recalculateVoteWeights(_accounts, voters) {
+		const accounts = _.cloneDeep(_accounts);
+
+		// Reset vote for all accounts
+		_.each(accounts, account => {
+			account.vote = '0';
+		});
+
+		// Recalculate vote
+		_.each(voters, delegate => {
+			let votes = '0';
+			const found = _.find(accounts, {
+				publicKey: Buffer.from(delegate.dependentId, 'hex'),
+			});
+			expect(found).to.be.an('object');
+			_.each(delegate.array_agg, voter => {
+				const found = _.find(accounts, {
+					address: voter,
+				});
+				votes = new Bignum(votes).plus(new Bignum(found.balance)).toString();
+			});
+			found.vote = votes;
+		});
+
+		return accounts;
+	}
+
+	function applyOutsiders(_accounts, delegatesList, blocks) {
+		const accounts = _.cloneDeep(_accounts);
+
+		_.each(delegatesList, publicKey => {
+			const found = _.find(blocks, {
+				generatorPublicKey: Buffer.from(publicKey, 'hex'),
+			});
+			const account = _.find(accounts, {
+				publicKey: Buffer.from(publicKey, 'hex'),
+			});
+			if (!found) {
+				account.missedBlocks += 1;
+			}
+		});
+
+		return accounts;
+	}
+
+	function getExpectedRoundRewards(blocks) {
+		const rewards = {};
+
+		const feesTotal = _.reduce(
+			blocks,
+			(fees, block) => {
+				return new Bignum(fees).plus(block.totalFee);
+			},
+			0
+		);
+
+		const rewardsTotal = _.reduce(
+			blocks,
+			(reward, block) => {
+				return new Bignum(reward).plus(block.reward);
+			},
+			0
+		);
+
+		const feesPerDelegate = new Bignum(feesTotal.toPrecision(15))
+			.dividedBy(slots.delegates)
+			.floor();
+		const feesRemaining = new Bignum(feesTotal.toPrecision(15)).minus(
+			feesPerDelegate.times(slots.delegates)
+		);
+
+		__testContext.debug(
+			`	Total fees: ${feesTotal.toString()} Fees per delegates: ${feesPerDelegate.toString()} Remaining fees: ${feesRemaining} Total rewards: ${rewardsTotal}`
+		);
+
+		_.each(blocks, (block, index) => {
+			const publicKey = block.generatorPublicKey.toString('hex');
+			if (rewards[publicKey]) {
+				rewards[publicKey].fees = rewards[publicKey].fees.plus(feesPerDelegate);
+				rewards[publicKey].rewards = rewards[publicKey].rewards.plus(
+					block.reward
+				);
+			} else {
+				rewards[publicKey] = {
+					publicKey,
+					fees: new Bignum(feesPerDelegate),
+					rewards: new Bignum(block.reward),
+				};
+			}
+
+			if (index === blocks.length - 1) {
+				// Apply remaining fees to last delegate
+				rewards[publicKey].fees = rewards[publicKey].fees.plus(feesRemaining);
```

Comment:
The `publicKey` here represents the block.generatorPublicKey for last block in collection. Is that sure the last block in the collection must be from last delegate in the round?

Reply:
Yep.